src/SA/ldapsearch/entry.c

```diff
@@ -0,0 +1,398 @@
+
+    internal_printf("%S %S", DateBuffer, TimeBuffer);
+}
+
+PCHAR GetDomainControllerHostName(){
+	DWORD dwRet;
+	PDOMAIN_CONTROLLER_INFO pdcInfo;
+
+	dwRet = NETAPI32$DsGetDcNameA(NULL, NULL, NULL, NULL, 0, &pdcInfo);
+	if (ERROR_SUCCESS == dwRet) {
+		internal_printf("[*] Using DC: %s\n", pdcInfo->DomainControllerName);		
+	} else {
+ 	    BeaconPrintf(CALLBACK_ERROR, "Failed to identify PDC, are we domain joined?");
+	}
+
+	PCHAR hostName = pdcInfo->DomainControllerName + 2; // Remove preceeding backslashes
+
+	NETAPI32$NetApiBufferFree(pdcInfo);
+	return hostName;
+}
+
+LDAP* InitialiseLDAPConnection(PCHAR hostName, PCHAR distinguishedName){
+	LDAP* pLdapConnection = NULL;
+
+    pLdapConnection = WLDAP32$ldap_init(hostName, 389);
+    
+    if (pLdapConnection == NULL)
+    {
+      	BeaconPrintf(CALLBACK_ERROR, "Failed to establish LDAP connection on 389.");
+        WLDAP32$ldap_unbind(pLdapConnection);
+    }
+    else {
+		internal_printf("[+] Initialised Connection.\n");
+    }
+
+	//////////////////////////////
+	// Bind to DC
+	//////////////////////////////
+    ULONG lRtn = 0;
+
+    lRtn = WLDAP32$ldap_bind_s(
+                pLdapConnection,      // Session Handle
+                distinguishedName,                // Domain DN
+                NULL,                 // Credential structure
+                LDAP_AUTH_NEGOTIATE); // Auth mode
+
+    if(lRtn == LDAP_SUCCESS)
+    {
+    	internal_printf("[+] Bind Successful.\n");
+    }
+    else
+    {
+      	BeaconPrintf(CALLBACK_ERROR, "Bind Failed: %i", lRtn);
+        WLDAP32$ldap_unbind(pLdapConnection);
+    }
+    return pLdapConnection;
+}
+
+LDAPMessage* ExecuteLDAPQuery(LDAP* pLdapConnection, PCHAR distinguishedName, char * ldap_filter, char * ldap_attributes){
+	ULONG errorCode = LDAP_SUCCESS;
+    LDAPMessage* pSearchResult;
+    
+	if(strlen(ldap_attributes)!=0){
+		PCHAR attr[1];
+		attr[0] = ldap_attributes;
+
+		errorCode = WLDAP32$ldap_search_s(
+        pLdapConnection,    // Session handle
+        distinguishedName,           // DN to start search
+        LDAP_SCOPE_SUBTREE, // Scope
+        ldap_filter,          // Filter
+        attr,      // Retrieve list of attributes
+        0,                  // Get both attributes and values
+        &pSearchResult);    // [out] Search results
+	} else {
+		errorCode = WLDAP32$ldap_search_s(
+        pLdapConnection,    // Session handle
+        distinguishedName,           // DN to start search
+        LDAP_SCOPE_SUBTREE, // Scope
+        ldap_filter,          // Filter
+        NULL,      // Retrieve list of attributes
+        0,                  // Get both attributes and values
+        &pSearchResult);    // [out] Search results
+	}   
+    
+    if (errorCode != LDAP_SUCCESS)
+    {
+        BeaconPrintf(CALLBACK_ERROR, "LDAP Search Failed: %i", errorCode);
+
+        WLDAP32$ldap_unbind_s(pLdapConnection);
+        if(pSearchResult != NULL)
+            WLDAP32$ldap_msgfree(pSearchResult);
```

Comment:
don't forget to initialize pSearchResult to null in your declaration on line 146.  depending on compiler if ldap_search_s doesn't set it to null on error this could cause a crash